crates/storage/provider/src/post_state/mod.rs

```diff
@@ -282,22 +299,65 @@ impl PostState {
     ///
     /// The reverted changes are removed from this post-state, and their effects are reverted.
     pub fn revert_to(&mut self, target_block_number: BlockNumber) {
-        let account_changes_to_revert = self.account_changes.drain_above(target_block_number);
-        for (_, accounts) in account_changes_to_revert.into_iter().rev() {
-            self.accounts.extend(accounts);
+        // Revert account state & changes
+        let removed_account_changes = self.account_changes.drain_above(target_block_number);
+        let changed_accounts = self
+            .account_changes
+            .iter()
+            .flat_map(|(_, account_changes)| account_changes.iter().map(|(address, _)| *address))
+            .collect::<BTreeSet<_>>();
+        let mut account_state: BTreeMap<Address, Option<Account>> = BTreeMap::default();
+        for address in changed_accounts {
+            let info = removed_account_changes
+                .iter()
+                .find_map(|(_, changes)| {
+                    changes.iter().find_map(|ch| (ch.0 == &address).then_some(*ch.1))
+                })
+                .unwrap_or(*self.accounts.get(&address).expect("exists"));
+            account_state.insert(address, info);
```

Comment:
Loop above tries to find the first changeset of reverted blocks to get old value or it gets values from "plain_state"